front/pages/w/[wId]/u/chat/[cId]/index.tsx

```diff
@@ -0,0 +1,1187 @@
+                    onClick={() => {
+                      setExpanded(true);
+                    }}
+                  />
+                )}
+              </div>
+            </>
+          )}
+          {!message.retrievals && <div className="">Loading...</div>}
+        </div>
+      </div>
+      {expanded && message.retrievals && (
+        <div className="ml-4 mt-2 flex flex-col space-y-1">
+          {message.retrievals.map((r, i) => {
+            return <DocumentView document={r} key={i} />;
+          })}
+        </div>
+      )}
+    </div>
+  ) : null;
+}
+
+export function MessageView({
+  user,
+  message,
+  loading,
+  isLatestRetrieval,
+}: {
+  user: UserType | null;
+  message: ChatMessageType;
+  loading: boolean;
+  isLatestRetrieval: boolean;
+}) {
+  return (
+    <div className="">
+      {message.role === "retrieval" ? (
+        <div className="flex flex-row">
+          <RetrievalsView message={message} isLatest={isLatestRetrieval} />
+        </div>
+      ) : (
+        <div className="my-2 flex flex-row items-start">
+          <div
+            className={classNames(
+              "min-w-6 flex h-8 w-8 flex-initial rounded-md",
+              message.role === "assistant" ? "bg-gray-50" : "bg-gray-0"
+            )}
+          >
+            {message.role === "assistant" ? (
+              <div className="flex scale-50 pl-2">
+                <PulseLogo animated={loading}></PulseLogo>
+              </div>
+            ) : (
+              <div className="flex">
+                <img
+                  className="h-8 w-8 rounded-md"
+                  src={user?.image || "https://gravatar.com/avatar/anonymous"}
+                  alt=""
+                />
+              </div>
+            )}
+          </div>
+          <div
+            className={classNames(
+              "ml-2 mt-1 flex flex-1 flex-col whitespace-pre-wrap",
+              message.role === "user" ? "italic text-gray-500" : "text-gray-700"
+            )}
+          >
+            {message.message}
+          </div>
+        </div>
+      )}
+    </div>
+  );
+}
+
+const COMMANDS: { cmd: string; description: string }[] = [
+  {
+    cmd: "/new",
+    description: "Start a new conversation",
+  },
+  {
+    cmd: "/follow-up",
+    description:
+      "Follow-up with the assistant without performing a document retrieval",
+  },
+  {
+    cmd: "/retrieve",
+    description: "Perform a document retrieval without querying the assistant",
+  },
+];
+
+export default function AppChat({
+  user,
+  owner,
+  workspaceDataSources,
+  prodCredentials,
+  chatSession,
+  gaTrackingId,
+}: InferGetServerSidePropsType<typeof getServerSideProps>) {
+  const router = useRouter();
+
+  const [isMac, setIsMac] = useState<boolean>(false);
+
+  useEffect(() => {
+    setIsMac(
+      typeof window !== "undefined"
+        ? navigator.userAgent.toUpperCase().indexOf("MAC") >= 0
+        : false
+    );
+  });
```

Comment:
You need an empty dependency array here otherwise it will be re-computed for each call to render() by React.

Reply:
:+1: